src/Workspaces/Core/Portable/PatternMatching/PatternMatch.cs

```diff
@@ -57,44 +58,16 @@ public int CompareTo(PatternMatch other)
             => CompareTo(other, ignoreCase: false);
 
         public int CompareTo(PatternMatch other, bool ignoreCase)
-        {
-            int diff;
-            if ((diff = CompareType(this, other)) != 0 ||
-                (diff = CompareCase(this, other, ignoreCase)) != 0 ||
-                (diff = ComparePunctuation(this, other)) != 0)
-            {
-                return diff;
-            }
-
-            return 0;
-        }
-
-        private static int ComparePunctuation(PatternMatch result1, PatternMatch result2)
-        {
-            // Consider a match to be better if it was successful without stripping punctuation
-            // versus a match that had to strip punctuation to succeed.
-            if (result1._punctuationStripped != result2._punctuationStripped)
-            {
-                return result1._punctuationStripped ? 1 : -1;
-            }
-
-            return 0;
-        }
-
-        private static int CompareCase(PatternMatch result1, PatternMatch result2, bool ignoreCase)
-        {
-            if (!ignoreCase)
-            {
-                if (result1.IsCaseSensitive != result2.IsCaseSensitive)
-                {
-                    return result1.IsCaseSensitive ? -1 : 1;
-                }
-            }
-
-            return 0;
-        }
-
-        private static int CompareType(PatternMatch result1, PatternMatch result2)
-            => result1.Kind - result2.Kind;
+            => ComparerWithState.CompareTo(this, other, ignoreCase, s_comparers);
+
+        private readonly static ImmutableArray<ComparerWithState<PatternMatch, bool>> s_comparers =
+            ComparerWithState.CreateComparers<PatternMatch, bool>(
+                // Compare types
+                (p, b) => p.Kind,
+                // Compare cases
+                (p, b) => !b && !p.IsCaseSensitive,
+                // Consider a match to be better if it was successful without stripping punctuation
+                // versus a match that had to strip punctuation to succeed.
+                (p, b) => p._punctuationStripped);
```

Comment:
fwiw, i'm reviewing this but not actually confirming the lambdas match the original code.   i hope you've done that.